src/NHibernate.Test/NHSpecificTest/DtcFailures/DtcFailuresFixture.cs

```diff
@@ -17,247 +17,308 @@ namespace NHibernate.Test.NHSpecificTest.DtcFailures
+					// Ensure the connection is acquired (thus enlisted)
+					Assert.DoesNotThrow(s.Flush, "Failure enlisting a connection in a distributed transaction.");
+				}
 			}
-			catch (AssertionException)
+		}
+
+		[Test]
+		public void SupportsPromotingToDistributed()
+		{
+			using (new TransactionScope())
 			{
-				throw;
+				using (var s = sessions.OpenSession())
+				{
+					s.Save(new Person { CreatedAt = DateTime.Now });
+					// Ensure the connection is acquired (thus enlisted)
+					s.Flush();
+				}
+				Assert.DoesNotThrow(() => ForceEscalationToDistributedTx.Escalate(),
+					"Failure promoting the transaction to distributed while already having enlisted a connection.");
+				Assert.AreNotEqual(
+					Guid.Empty,
+					System.Transactions.Transaction.Current.TransactionInformation.DistributedIdentifier,
+					"Transaction lacks a distributed identifier");
 			}
-			catch (Exception) {}
 		}
 
 		[Test]
-		public void Can_roll_back_transaction()
+		public void WillNotCrashOnDtcPrepareFailure()
 		{
-			if (Dialect is FirebirdDialect)
-				Assert.Ignore("Firebird driver does not support distributed transactions");
-
 			var tx = new TransactionScope();
-			using (ISession s = sessions.OpenSession())
+			var disposeCalled = false;
+			try
 			{
-				new ForceEscalationToDistributedTx(true); //will rollback tx
-				s.Save(new Person { CreatedAt = DateTime.Today });
+				using (var s = sessions.OpenSession())
+				{
+					s.Save(new Person { NotNullData = null }); // Cause a SQL not null constraint violation.
+				}
+
+				ForceEscalationToDistributedTx.Escalate();
 
 				tx.Complete();
+				disposeCalled = true;
+				Assert.Throws<TransactionAbortedException>(tx.Dispose, "Scope disposal has not rollback and throw.");
 			}
+			finally
+			{
+				if (!disposeCalled)
+				{
+					try
+					{
+						tx.Dispose();
+					}
+					catch
+					{
+						// Ignore, if disposed has not been called, another exception has occurred in the try and
+						// we should avoid overriding it by the disposal failure.
+					}
+				}
+			}
+		}
+
+		[Test]
+		public void Can_roll_back_transaction()
+		{
+			var tx = new TransactionScope();
+			var disposeCalled = false;
 			try
 			{
-				tx.Dispose();
-				Assert.Fail("Expected tx abort");
+				using (var s = sessions.OpenSession())
+				{
+					ForceEscalationToDistributedTx.Escalate(true); //will rollback tx
+					s.Save(new Person { CreatedAt = DateTime.Today });
+
+					tx.Complete();
+				}
+				disposeCalled = true;
+				Assert.Throws<TransactionAbortedException>(tx.Dispose, "Scope disposal has not rollback and throw.");
 			}
-			catch (TransactionAbortedException)
+			finally
 			{
-				//expected   
+				if (!disposeCalled)
+				{
+					try
+					{
+						tx.Dispose();
+					}
+					catch
+					{
+						// Ignore, if disposed has not been called, another exception has occurred in the try and
+						// we should avoid overriding it by the disposal failure.
+					}
+				}
 			}
+
+			AssertNoPersons();
```

Comment:
Check the save has correctly been rollback-ed. Not the case with Firebird by example.